jobsdb/jobsdb.go

```diff
@@ -1615,26 +1625,55 @@ func (jd *HandleT) invalidateCache(ds dataSetT) {
 	}
 }
 
-//Rename a dataset
-func (jd *HandleT) renameDS(ds dataSetT, allowMissing bool) {
+//mustRenameDS renames a dataset
+func (jd *HandleT) mustRenameDS(ds dataSetT) {
 	var sqlStatement string
 	var renamedJobStatusTable = fmt.Sprintf(`pre_drop_%s`, ds.JobStatusTable)
 	var renamedJobTable = fmt.Sprintf(`pre_drop_%s`, ds.JobTable)
-
-	if allowMissing {
-		sqlStatement = fmt.Sprintf(`ALTER TABLE IF EXISTS "%s" RENAME TO "%s"`, ds.JobStatusTable, renamedJobStatusTable)
-	} else {
+	err := jd.doInTransaction(func(tx *sql.Tx) error {
 		sqlStatement = fmt.Sprintf(`ALTER TABLE "%s" RENAME TO "%s"`, ds.JobStatusTable, renamedJobStatusTable)
-	}
-	_, err := jd.dbHandle.Exec(sqlStatement)
+		_, err := jd.dbHandle.Exec(sqlStatement)
+		if err != nil {
+			return err
+		}
+
+		sqlStatement = fmt.Sprintf(`ALTER TABLE "%s" RENAME TO "%s"`, ds.JobTable, renamedJobTable)
+		_, err = jd.dbHandle.Exec(sqlStatement)
+		if err != nil {
+			return err
+		}
+		if len(jd.preBackupHandlers) > 0 {
+			for _, preBackupHandler := range jd.preBackupHandlers {
+				err = preBackupHandler.Handle(context.TODO(), tx, renamedJobTable, renamedJobStatusTable)
+				if err != nil {
+					return err
+				}
+			}
+		}
+		return nil
+	})
 	jd.assertError(err)
+}
+
+// renameDS renames a dataset if it exists
+func (jd *HandleT) renameDS(ds dataSetT) {
+	var sqlStatement string
+	var renamedJobStatusTable = fmt.Sprintf(`pre_drop_%s`, ds.JobStatusTable)
+	var renamedJobTable = fmt.Sprintf(`pre_drop_%s`, ds.JobTable)
+	err := jd.doInTransaction(func(tx *sql.Tx) error {
+		sqlStatement = fmt.Sprintf(`ALTER TABLE IF EXISTS "%s" RENAME TO "%s"`, ds.JobStatusTable, renamedJobStatusTable)
```

Comment:
Idr if it would work because we're talking about a table name and not a regular argument but perhaps you know. Do you need `fmt.Sprintf` here or we can use prepared statements instead? As in doing `ALTER TABLE $1` and then `db.dbHandle.Exec(sqlStatement, "TABLE_NAME")`.

Reply:
Yeah it's not possible, it looks like it could come in the future though: https://github.com/golang/go/issues/18478

However, something we could do is use [QuoteIdentifier](https://pkg.go.dev/github.com/lib/pq#QuoteIdentifier) instead. Potentially we could also avoid `fmt.Sprintf` which uses reflection (slower), not really needed though.

Reply:
Yes & no.  Double-quotes & zero-bytes is not something that we are really at risk of, given the conventions that we are currently using for table names in rudder-server. I don't have any objection to doing what you propose though, as long as we do this on a separate task and for all occurrences in the codebase, otherwise it doesn't offer any advantage. Still I would treat it with low priority.

Reply:
Yeah sure, no problem with `fmt.Sprintf()` as well, it's a super minor optimization, I mostly asked because I genuinely didn't remember whether it was possible to "prepare" a table name :sweat_smile: